File: src/main/java/com/example/skyr/note/NoteDao.java
package com.example.skyr.note;

import com.example.skyr.Dao;
import com.example.skyr.InstantFactory;
import com.example.skyr.exception.ServiceErrorException;
import com.example.skyr.pagination.PaginatedResult;
import com.example.skyr.pagination.Pagination;
import lombok.RequiredArgsConstructor;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor
public final class NoteDao extends Dao {
    private final JdbcTemplate jdbcTemplate;

    public void create(final Note note) {
        create(note.id(), note.title(), note.content(), note.creationTime(), note.modificationTime());
    }

    public void create(final UUID id, final String title, final String content) {
        final var creationTime = InstantFactory.create();
        create(id, title, content, creationTime, creationTime);
    }

    public void create(final UUID id,
                       final String title,
                       final String content,
                       final Instant creationTime,
                       final Instant modificationTime) {
        jdbcTemplate.update(
                "INSERT INTO note VALUES (?, ?, ?, ?, ?)",
                id.toString(),
                title,
                content,
                timestamp(creationTime).toString(),
                timestamp(modificationTime).toString()
        );
    }

    public boolean delete(final UUID id) {
        final var result = jdbcTemplate.update("DELETE FROM note WHERE id = ?", id.toString());
        return isUpdated(result);
    }

    public Note read(final UUID id) {
        final var result = jdbcTemplate.query(
                "SELECT * FROM note WHERE id = ?",
                mapper(),
                id.toString()
        );
        return singleResult(result);
    }

    public List<Note> readWithLock(final Instant modificationTimeCursor) {
        return jdbcTemplate.query(
                "SELECT * FROM note WHERE modification_time > ? ORDER BY modification_time FOR UPDATE",
                mapper(),
                timestamp(modificationTimeCursor)
        );
    }

    public PaginatedResult<Note> read(final Pagination pagination) {
        final var result = jdbcTemplate.query(
                "SELECT * FROM note LIMIT ? OFFSET ?",
                mapper(),
                limit(pagination.pageSize()),
                offset(pagination.pageNumber(), pagination.pageSize())
        );
        return paginatedResult(pagination, result);
    }

    public void update(final Note note) {
        update(note.id(), note.title(), note.content(), note.creationTime(), note.modificationTime());
    }

    public void update(final UUID id,
                       final String title,
                       final String content,
                       final Instant creationTime,
                       final Instant modificationTime) {
        final var parameters = new LinkedHashMap<String, String>();
        if (title != null) parameters.put(Note.TITLE, title);
        if (content != null) parameters.put(Note.CONTENT, content);
        if (creationTime != null) parameters.put(Note.CREATION_TIME, timestamp(creationTime).toString());
        if (modificationTime != null) parameters.put(Note.MODIFICATION_TIME, timestamp(modificationTime).toString());
        if (parameters.isEmpty()) throw new ServiceErrorException("Update parameters cannot be empty");
        final var setters = parameters
                .keySet()
                .stream()
                .map(key -> key + " = ?")
                .collect(Collectors.joining(", "));
        parameters.put(Note.ID, id.toString());
        jdbcTemplate.update(
                "UPDATE note SET " + setters + " WHERE id = ?",
                parameters.values().toArray()
        );
    }

    private RowMapper<Note> mapper() {
        return (resultSet, rowNum) -> NoteFactory.create(resultSet);
    }
}

File: src/main/kotlin/com/example/skyr/note/NoteDao.kt
package com.example.skyr.note

import com.example.skyr.Dao
import com.example.skyr.InstantFactory
import com.example.skyr.pagination.PaginatedResult
import com.example.skyr.pagination.Pagination
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.RowMapper
import org.springframework.stereotype.Component
import java.time.Instant
import java.util.*

@Component
class NoteDao(private val jdbcTemplate: JdbcTemplate) : Dao() {

    companion object {
        private const val NOTE = "note"
    }

    fun create(note: Note) {
        create(note.id, note.title, note.content, note.creationTime, note.modificationTime)
    }

    fun create(id: UUID, title: String, content: String) {
        val creationTime = InstantFactory.create()
        create(id, title, content, creationTime, creationTime)
    }

    fun create(id: UUID, title: String, content: String, creationTime: Instant, modificationTime: Instant) {
        jdbcTemplate.update(
            "INSERT INTO $NOTE VALUES (?, ?, ?, ?, ?)",
            id.toString(),
            title,
            content,
            timestamp(creationTime).toString(),
            timestamp(modificationTime).toString()
        )
    }

    fun delete(id: UUID): Boolean {
        val result = jdbcTemplate.update("DELETE FROM $NOTE WHERE ${Note.ID} = ?", id.toString())
        return resolveResult(result)
    }

    fun read(id: UUID): Note? {
        val result = jdbcTemplate.query(
            "SELECT * FROM $NOTE WHERE ${Note.ID} = ?",
            mapper(),
            id.toString()
        )
        return resolveSingleResult(result)
    }

    fun read(modificationTimeCursor: Instant, writeLock: Boolean = false): List<Note> {
        return jdbcTemplate.query(
            "SELECT * FROM $NOTE WHERE ${Note.MODIFICATION_TIME} > ? ORDER BY ${Note.MODIFICATION_TIME}" + forUpdate(writeLock),
            mapper(),
            timestamp(modificationTimeCursor)
        )
    }

    fun read(pagination: Pagination): PaginatedResult<Note> {
        val result = jdbcTemplate.query(
            "SELECT * FROM $NOTE LIMIT ? OFFSET ?",
            mapper(),
            limit(pagination.pageSize),
            offset(pagination.pageNumber, pagination.pageSize)
        )
        return resolvePaginatedResult(pagination, result)
    }

    fun update(note: Note) {
        update(note.id, note.title, note.content, note.creationTime, note.modificationTime)
    }

    fun update(
        id: UUID,
        title: String? = null,
        content: String? = null,
        creationTime: Instant? = null,
        modificationTime: Instant? = null
    ) {
        var parameters =
            listOfNotNull(title, content, timestamp(creationTime)?.toString(), timestamp(modificationTime)?.toString())
        if (parameters.isEmpty()) return
        parameters = parameters.toMutableList()
        parameters.add(id.toString())
        val setters = listOfNotNull(
            if (title == null) null else Note.TITLE,
            if (content == null) null else Note.CONTENT,
            if (creationTime == null) null else Note.CREATION_TIME,
            if (modificationTime == null) null else Note.MODIFICATION_TIME,
        ).stream().map { setter -> "$setter = ?" }.toList()
        jdbcTemplate.update(
            "UPDATE $NOTE SET ${setters.joinToString()} WHERE ${Note.ID} = ?",
            *(parameters.toTypedArray())
        )
    }

    private fun mapper(): RowMapper<Note> {
        return RowMapper { resultSet, _ -> note(resultSet) }
    }
}
